### PR TITLE
filtering a list

### LINKED ISSUE #28: filtering a list

### PR DESCRIPTION
**Pull request by HafizHussein**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/filter_list.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+# --- before documenting ---
+# def filter_list(lst):
+#'return a new list with the strings filtered out'
+
+
+# --- after documenting ---
+
+
+def filter_list(lst):
```
Comment (sky0-0glitch) [inline solutions/filter_list.py]: add type annotations to your function

File: solutions/filter_list.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    """
+    Filters out all strings from a list, keeping only non-negative integers.
+
+    Parameters:
+        lst: list, a list containing non-negative integers and strings
+
+    Returns -> list: a new list with strings removed
+
```
Comment (sky0-0glitch) [inline solutions/filter_list.py:37]: Add defensive assertions using Raises

**Review by sky0-0glitch — CHANGES_REQUESTED** [review]

File: solutions/filter_list.py (hunk 1 above)
Comment (HafizHussein) [reply solutions/filter_list.py]: Added.

File: solutions/filter_list.py (hunk 2 above)
Comment (HafizHussein) [reply solutions/filter_list.py:37]: Added.

File: solutions/filter_list.py (hunk 1 above)
Comment (sky0-0glitch) [reply solutions/filter_list.py]: hello dear your function type annotation should be something like: def filter_list(lst: list) -> list
you are doing a great job🙌

Comment (HafizHussein) [reply solutions/filter_list.py]: Thank you again, added the annotation.
You are also doing an amazing job!

Comment (sky0-0glitch) [reply solutions/filter_list.py]: Oh you commented that but you have to write in line 29 like what you did in the substracting numbers challenge.

Comment (HafizHussein) [reply solutions/filter_list.py]: I did already

Comment (sky0-0glitch) [reply solutions/filter_list.py]: Screenshot 2025-01-05 000046
 this is what I  am seeing while it should have been this `def filter_list(lst: list) -> list` are you sure you have pushed all your changes?

Comment (HafizHussein) [reply solutions/filter_list.py]: my bad did it in line 22 but fixed it now.
hopefully you see it by your end.

Comment (sky0-0glitch) [reply solutions/filter_list.py]: yep I can see it now which is why I deleted my comment but I was late😂. I think we are done here great job 🙌

**Review by sky0-0glitch — APPROVED** [review]
